fields/src/macros.rs

```diff
@@ -86,55 +86,81 @@ macro_rules! impl_primefield_from_int {
+                let g = $Self::two_adic_root_of_unity().pow($P::T);
+                let two = $Self::from(2u8);
+                let n_field = $Self::from(n);
+
+                let k = ((n - 1) as f64).sqrt().floor() as usize;
+                let k_field = $Self::from(k as u64);
+                let k_2 = ((k as f64) / 2.0).floor() as usize;
+                let k_2_field = $Self::from(k_2 as u64);
+                let k_1 = k - k_2;
+                let l_minus_one_times_k = n_field - $Self::one() - k_2_field;
+                let l_minus_one = l_minus_one_times_k / k_field;
+                let l = l_minus_one + $Self::one();
+                let mut l_s: Vec<$Self> = Vec::with_capacity(k);
+
+                for _ in 0..k_1 {
+                    l_s.push(l_minus_one);
+                }
+                for _ in 0..k_2 {
+                    l_s.push(l);
                 }
 
-                while !b.is_one() {
-                    let mut k = 0usize;
+                let mut x_s: Vec<$Self> = Vec::with_capacity(k);
+                for i in 0..k {
+                    x_s.push(x.pow(
+                        two.pow((n_field - $Self::one() - (l_s.iter().take(i + 1).sum::<$Self>())).to_repr()).to_repr(),
+                    ));
+                }
 
-                    let mut b2k = b;
-                    while !b2k.is_one() {
-                        // invariant: b2k = b^(2^k) after entering this loop
-                        b2k.square_in_place();
-                        k += 1;
+                let find = |delta: $Self| -> $Self {
+                    let mut mu = delta;
+                    let mut i = $Self::zero();
+                    while mu != -$Self::one() {
+                        mu.square_in_place();
+                        i += $Self::one();
                     }
-
-                    let j = v - k - 1;
-                    w = z;
-                    for _ in 0..j {
-                        w.square_in_place();
+                    i
+                };
+
+                let eval = |alpha: $Self| -> $Self {
+                    let mut delta = alpha;
+                    let mut s = $Self::zero();
+                    while delta != $Self::one() {
+                        let i = find(delta);
+                        let n_minus_one_minus_i = (n_field - $Self::one() - i).to_repr();
+                        s += two.pow(n_minus_one_minus_i);
+                        if i > $Self::zero() {
+                            delta *= g.pow(two.pow(n_minus_one_minus_i).to_repr());
+                        } else {
+                            delta = -delta;
+                        }
                     }
-
-                    z = w.square();
-                    b *= &z;
-                    x *= &w;
-                    v = k;
+                    s
+                };
+
+                let mut s = $Self::zero();
+                let mut t = $Self::zero();
+                for i in 0..k {
```

Comment:
it would be best to work on a `l_s.iter().zip(x_s.iter())` here to remove index checking